lib/services/label_expressions.go

```diff
@@ -139,3 +142,51 @@ func newLabelExpressionParser() (*typical.Parser[labelExpressionEnv, bool], erro
 	})
 	return parser, trace.Wrap(err)
 }
+
+// labelsMatching returns the aggregate of all label values for all keys that
+// match keyExpr. It supports globs or full regular expressions and must find
+// a complete match for the key.
+func labelsMatching(env labelExpressionEnv, keyExpr string) ([]string, error) {
+	allLabels := env.resourceLabelGetter.GetAllLabels()
+	var matchingLabelValues []string
+	for key, value := range allLabels {
+		match, err := utils.MatchString(key, keyExpr)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		if match {
+			matchingLabelValues = append(matchingLabelValues, value)
+		}
+	}
+	return matchingLabelValues, nil
+}
+
+// containsAny returns true if [list] contains any element of [items].
+func containsAny(list []string, items []string) (bool, error) {
+	s := set(list)
+	for _, item := range items {
+		if _, ok := s[item]; ok {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+// containsAny returns true if [list] contains every element of [items].
+func containsAll(list []string, items []string) (bool, error) {
+	s := set(list)
+	for _, item := range items {
+		if _, ok := s[item]; !ok {
+			return false, nil
+		}
+	}
+	return true, nil
```

Comment:
Good point, I think it would better to avoid this, for example I think we wouldn't want an expression like `contains_all(user.spec.traits["projects"], labels_matching("project-*"))` to match for nodes that have no project label. I've updated this to return false in that case.